File: sample-objc/ARGEARSample/ARGearHelpers/NetworkManager.m
//
//  NetworkManager.m
//  ARGEARSample
//
//  Created by Jihye on 10/10/2019.
//  Copyright © 2019 Seerslab. All rights reserved.
//

#import "NetworkManager.h"
#import <AFNetworking/AFNetworking.h>

@implementation NetworkManager

+ (NetworkManager *)shared {
    static dispatch_once_t pred;
    static NetworkManager *sharedInstance = nil;
    dispatch_once(&pred, ^{
        sharedInstance = [[NetworkManager alloc] init];
    });
    return sharedInstance;
}

- (instancetype)init {
    self = [super init];
    if (self) {
    }
    return self;
}

- (void)connectAPI:(void (^_Nonnull)(BOOL success, NSDictionary * _Nullable data))completionHandler {
    
    NSString *apiUrlString = [NSString stringWithFormat:@"%@%@", API_HOST, API_KEY];
    NSURL *url = [NSURL URLWithString:apiUrlString];
    NSURLRequest *request = [NSURLRequest requestWithURL:url];
    
    NSURLSessionDataTask *dataTask = [[NSURLSession sharedSession] dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        if (error) {
            NSLog(@"connectAPI error %@", error);
        } else {
            NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)response;
            NSLog(@"response status code: %ld", (long)[httpResponse statusCode]);
            
            NSError *jsonError = nil;
            id jsonObject = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:&jsonError];
            
            if (jsonError) {
                completionHandler(NO, nil);
                return;
            }

            if ([jsonObject isKindOfClass:[NSDictionary class]]) {
                completionHandler(YES, (NSDictionary *)jsonObject);
            } else {
                completionHandler(NO, nil);
            }
        }
    }];
    
    [dataTask resume];
}

- (void)downloadItem:(Item *)item
             success:(void (^)(NSString *uuid, NSURL *targetPath))successBlock
                fail:(void (^)(void))failBlock {
    
    NSString *uuid = [item uuid];
    NSString *fileUrl = [item zip_file];
    NSString *type = [item type];
    NSString *title = [item title];
    
    ARGAuthCallback callback;
    
    callback.Success = ^(NSString *url) {
        
        NSURLSessionConfiguration *configuration = [NSURLSessionConfiguration defaultSessionConfiguration];
        AFURLSessionManager *manager = [[AFURLSessionManager alloc] initWithSessionConfiguration:configuration];

        NSURL *URL = [NSURL URLWithString:url];
        NSURLRequest *request = [NSURLRequest requestWithURL:URL];

        NSURLSessionDownloadTask *downloadTask = [manager downloadTaskWithRequest:request progress:nil destination:^NSURL *(NSURL *targetPath, NSURLResponse *response) {
            NSURL *cacheDirectoryURL = [[NSFileManager defaultManager] URLForDirectory:NSCachesDirectory
                                                                              inDomain:NSUserDomainMask
                                                                     appropriateForURL:nil
                                                                                create:NO
                                                                                 error:nil];

            return [cacheDirectoryURL URLByAppendingPathComponent:[response suggestedFilename]];
        } completionHandler:^(NSURLResponse *response, NSURL *filePath, NSError *error) {

            if (error) {
                failBlock();
                return;
            }

            successBlock(uuid, filePath);
        }];

        [downloadTask resume];
    };
    
    callback.Error = ^(ARGStatusCode code) {
        
        [self alertByCode:code];
        failBlock();
    };

    [[_session auth] requestSignedUrlWithUrl:fileUrl itemTitle:title itemType:type completion:callback];
}

- (void)alertByCode:(ARGStatusCode)code {
    
    NSString *errorTitle = @"";
    NSString *errorMessage = @"";
    
    switch (code) {
        case ARGStatusCode_NETWORK_OFFLINE:
            errorTitle = @"Network offline";
            break;
        case ARGStatusCode_NETWORK_TIMEOUT:
            errorTitle = @"Network timeout";
            break;
        case ARGStatusCode_NETWORK_ERROR:
            errorTitle = @"Network error";
            break;
        case ARGStatusCode_INVALID_AUTH:
            errorTitle = @"Invalid auth";
            break;
        default:
            break;
    }
    errorMessage = [NSString stringWithFormat:@"error code %ld", (long)code];
    
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:errorTitle message:errorMessage preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction *action) {
    }];
    [alertController addAction:okAction];

    [[[[UIApplication sharedApplication] keyWindow] rootViewController] presentViewController:alertController animated:YES completion:nil];
}

@end

File: sample-swift/ARGearSample/ARGearHelpers/Network/NetworkManager.swift
//
//  NetworkManager.swift
//  ARGearSample
//
//  Created by Jihye on 2020/01/07.
//  Copyright © 2020 Seerslab. All rights reserved.
//

import Foundation
import ARGear

let API_HOST = "https://apis.argear.io/api/v3/"
let API_KEY = "48cdd456621ebddb5283f516"
let API_SECRET_KEY = "e0ea29ce12178111d2a5e6093a6e048a2a6248b921f092f68c93992ca29a45ee"
let API_AUTH_KEY = "U2FsdGVkX19Sm4NdLCm7kZRApYCcdTaQqY7RbgodU5GHd9plkoHb24uqdAv0MICHb1aldMXpCHccDDeCTWqq/A=="

enum APIError: Error {
    case network
    case data
    case serializeJSON
}

enum DownloadError: Error {
    case network
    case auth
    case content
}

class NetworkManager {

    static let shared = NetworkManager()
    
    var argSession: ARGSession?
    
    init() {
    }
    
    func connectAPI(completion: @escaping (Result<[String: Any], APIError>) -> Void) {
        
        let urlString = API_HOST + API_KEY
        let url = URL(string: urlString)!
        let task = URLSession.shared.dataTask(with: url) { data, response, error in
            if let _ = error {
                return completion(.failure(.network))
            } else {
                guard let data = data else {
                    return completion(.failure(.data))
                }
                
                guard let json: [String : Any] = try? JSONSerialization.jsonObject(with: data, options: []) as? [String : Any] else {
                    return completion(.failure(.serializeJSON))
                }
                
                completion(.success(json))
            }
        }
        task.resume()
    }
    
    func downloadItem(_ item: Item, completion: @escaping (Result<URL, DownloadError>) -> Void) {
        guard let session = self.argSession, let auth = session.auth
            else {
                completion(.failure(.auth))
                return
        }
        
        guard let zipUrl = item.zip_file
            else {
                completion(.failure(.content))
                return
        }

        let authCallback : ARGAuthCallback = {(url: String?, code: ARGStatusCode) in
            if (code.rawValue == ARGStatusCode.SUCCESS.rawValue) {
                guard let url = url
                    else {
                        completion(.failure(.auth))
                        return
                }
                
                // download task
                let authUrl = URL(string: url)!
                let task = URLSession.shared.downloadTask(with: authUrl) { (downloadUrl, response, error) in
                   if error != nil {
                       completion(.failure(.network))
                       return
                   }

                   guard
                       let httpResponse = response as? HTTPURLResponse,
                       let response = response,
                       let downloadUrl = downloadUrl
                       else {
                           completion(.failure(.network))
                           return
                   }

                   if httpResponse.statusCode == 200 {
                       guard
                           var cachesDirectory = FileManager.default.urls(for: .cachesDirectory, in: .allDomainsMask).first,
                           let suggestedFilename = response.suggestedFilename
                           else {
                               completion(.failure(.content))
                               return
                       }
                       cachesDirectory.appendPathComponent(suggestedFilename)

                       let fileManager = FileManager.default
                       // remove
                       do {
                           try fileManager.removeItem(at: cachesDirectory)
                       } catch {
                       }
                       // copy
                       do {
                           try fileManager.copyItem(at: downloadUrl, to: cachesDirectory)
                       } catch {
                           completion(.failure(.content))
                           return
                       }

                       completion(.success(cachesDirectory))
                       return
                   }
                   completion(.failure(.network))
                }
                task.resume()
            } else {
                if code.rawValue > ARGStatusCode.VALID_AUTH.rawValue {
                    completion(.failure(.auth))
                } else {
                    completion(.failure(.network))
                }
            }
        }

        auth.requestSignedUrl(withUrl: zipUrl, itemTitle: item.title ?? "", itemType: item.type ?? "", completion: authCallback)
    }
}
